### PR TITLE
docs: document how to install custom tree-sitter grammars

### PR DESCRIPTION
**Pull request by menixator**
Adds a section to tree-sitter's documentation describing how to install your own grammars.

### CODE REVIEW

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:48]: this is a deprecated option, we should not be using it as an example. This is already a large section and it's not related to the functionality being documented, so it might make sense to just remove, as well. 
```suggestion
          settings.indent.enable = true;
```

File: plugins/languages/treesitter/treesitter.nix
Comment (MattSturgeon) [inline plugins/languages/treesitter/treesitter.nix:18]: Will need to check using our `serve-docs` devshell command, but I have a feeling mdbook doesn't support markdown headers within the option descriptions. The plugin description may work differently though, not sure.

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:62]: Did you test this out? From my quick experimenting and following their [readme](https://github.com/nvim-treesitter/nvim-treesitter/?tab=readme-ov-file#adding-parsers) this path needs to find the files below but you link to the derivation above.

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:62]: Yes. These two examples are actually from my config.

Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix:62]: This section is used for `:TSInstall nu`. But, you install it below with `extraPlugins`. If you ran `:TSInstall nu` you should see a message about a parser not managed by treesitter.

File: plugins/languages/treesitter/treesitter.nix
Comment (MattSturgeon) [inline plugins/languages/treesitter/treesitter.nix:58]: Nit: I find examples are easier to read with explicit parens on function calls.
```suggestion
            local parser_config = require("nvim-treesitter.parsers").get_parser_configs()
```

(I haven't checked if the same change is needed elsewhere in the example)

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix:62]: I was able follow the example and set up [fsharp](https://github.com/khaneliman/astronvim/commit/71227a14435ce93997b3e79a04d8a8b761cfe6fa). But, in my own testing of the installation, I didn't find the `extraConfigLua` part to be necessary when using `grammarPackages` with  `nixGrammars`. That section would be relevant for user's not installing through `nixGrammars` and utilizing the `settings.ensure_installed`.

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:62]: Hmm. This might be because I'm still on the `24.05` branch. I could not get it to work without adding it into the runtime path like that.

File: plugins/languages/treesitter/treesitter.nix
Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix]: This was the only instance that I could find.

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix:18]: image

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:45]: we should collapse all of these into a nested structure. 
```
programs.nixvim = {
  extraPlugins = [
  ...
  plugins = {
  ...
  extraConfigLua ==
```

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:27]:
```suggestion
      # example of building your own grammar:
```

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:74]:
```suggestion
      # Add as extra plugins so that their `queries/*.scm` get installed and can be picked up by `treesitter`. 
```

File: plugins/languages/treesitter/treesitter.nix
Comment (MattSturgeon) [reply plugins/languages/treesitter/treesitter.nix:18]: Awesome, thanks for checking! We could debate which heading level to use, but I don't want to get too nit-picky! (I'd probably prefer `###` or `####`, but wouldn't like to say without checking how it renders).

Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix:18]: Looks better with `###`. One thing I'm noticing is that it gets rid of the plugin doc link and maintainer link with a description. We might want to fix that.
image

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:18]:
```suggestion
    ### Installing Your Own Grammars with Nixvim
```

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:16]:
```suggestion
```

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix:18]: >  One thing I'm noticing is that it gets rid of the plugin doc link and maintainer link with a description. We might want to fix that.

Nvm, I'll create a PR. I see what it's doing.

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix]: @menixator Sorry to flip on this, I didn't realize it would hotlink to actual tree-sitter and not the plugin again. I think it's safe to add this back.

Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix]: Afterwards, can you squash all the commits into a single one.

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:16]: @khaneliman done.

File: plugins/languages/treesitter/treesitter.nix
Comment (MattSturgeon) [inline plugins/languages/treesitter/treesitter.nix:22]: I think this is confusingly worded. `pkgs.vimPlugins.nvim-treesitter` is a derivation (all packages are), but the `plugins.treesitter` module is _not_ a derivation.

I'm not sure it's strictly accurate to say the package derivation generates the grammar derivations, since that is done in this module by providing an overridden package using the `withPlugins` function.

Not sure the best way to word it though.

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:22]: How about:
```
If the grammar you want to install isn't provided through `nixpkgs`, you could install it like so:
```

Comment (MattSturgeon) [reply plugins/languages/treesitter/treesitter.nix:22]: I think it's nice to keep an implication that most users don't need to do anything, and that configuring `plugins.treesitter.nixGrammars` and/or `plugins.treesitter.grammarPackages` should cover 80% of usage when users _do_ need to do something.

> If the grammar you want to install isn't available through `nixpkgs`, you could install it like so:

But something like that could definitely fit into the text!

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:22]: > I think it's nice to keep an implication that most users don't need to do anything

If that's the case, these docs might be a better fit in some other document since it is sort of a niche case. Having it front and centre of such a ubiquitous plugin might scare people off 👀

Comment (MattSturgeon) [reply plugins/languages/treesitter/treesitter.nix:22]: I think it's fine to have it here. Let's not let perfect be the enemy of good!

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix]: I've reworded it. Let me know if that works for you.

File: plugins/languages/treesitter/treesitter.nix
Comment (khaneliman) [inline plugins/languages/treesitter/treesitter.nix:21]:
```suggestion
    But, in the rare case it isn't, you can build your own and use it with Nixvim
```

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
One suggestion I'd like addressed, and a couple _very_ minor nits which you're welcome to ignore.

When you're done are you also able to squash and rebase onto our `main` branch? No big deal if not as we can handle that.

Thanks for working on this, documentation is far too often neglected! :heart: 


@khaneliman what do you think about a follow up PR where we (somehow) include the examples in the test suite? E.g. they could be `readFile`-ed into the description and also added to the test-sources in a similar way to how `examples/` are.

**Comment by khaneliman** [comment]
> @khaneliman what do you think about a follow up PR where we (somehow) include the examples in the test suite? E.g. they could be `readFile`-ed into the description and also added to the test-sources in a similar way to how `examples/` are.

I think it could be pretty useful if we filter out the default config so we don't double up on those examples in documentation. Could probably use a special attribute name designation in tests to be included in the description.

File: plugins/languages/treesitter/treesitter.nix
Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:88]: Oh that's neovim's `gq` doing its thing haha. Let me just do it manually.

I should really add in something like http://www.nicemice.net/par/ or https://github.com/wincent/wincent/blob/main/aspects/nvim/files/.config/nvim/plugin/par.lua  to neovim to do that. Let me just do this manually.

**Comment by menixator** [comment]
> When you're done are you also able to squash and rebase onto our `main` branch? No big deal if not as we can handle that.

Done. 😄

File: plugins/languages/treesitter/treesitter.nix
Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix]: Fixed now.

**Review by MattSturgeon — APPROVED** [review]
Awesome, LGTM thanks!

I'll let @khaneliman have the final say since he handled most of the review.

**Review by khaneliman — APPROVED** [review]
LGTM, thanks for your help documenting how to use the module!

**Comment by khaneliman** [comment]
@mergifyio queue

File: plugins/languages/treesitter/treesitter.nix
Comment (menixator) [inline plugins/languages/treesitter/treesitter.nix:85]: @MattSturgeon / @khaneliman    the ```nix should be moved to a newline here. It borked the documentation completely haha. 
https://nix-community.github.io/nixvim/plugins/treesitter/index.html?highlight=tree-sitter#installing-your-own-grammars-with-nixvim

Comment (khaneliman) [reply plugins/languages/treesitter/treesitter.nix:85]: Doh, if you wanna do a quick PR I can approve and merge

Comment (menixator) [reply plugins/languages/treesitter/treesitter.nix:85]: Done #2073

**Pull request merged**
